src/stan/model/model_base_crtp.hpp

```diff
@@ -0,0 +1,216 @@
+#ifndef STAN_MODEL_MODEL_BASE_CRTP_HPP
+#define STAN_MODEL_MODEL_BASE_CRTP_HPP
+
+#include <stan/model/model_base.hpp>
+#include <iostream>
+#include <utility>
+#include <vector>
+
+namespace stan {
+namespace model {
+
+/**
+ * Base class employing the curiously recursive template pattern for
+ * static inheritance to adapt templated `log_prob` and `write_array`
+ * methods to their untemplated virtual counterparts declared in
+ * `model_base`.
+ *
+ * The derived class `M` is required to implement the following two
+ * pairs of template functions,
+ *
+ * ```
+ * template <bool propto, bool jacobian, typename T>
+ * T log_prob(std::vector<T>& params_r,
+ *              std::vector<int>& params_i,
```

Comment:
minor: mind lining up the `std::vector`s? (remove two spaces from L24 and L25)

<img width="631" alt="Untitled" src="https://user-images.githubusercontent.com/425751/61265414-0ab66200-a75e-11e9-8f91-5faf0e7c3a33.png">

Reply:
Done.